### ISSUE TITLE
3D Reacher Training

### ISSUE THREAD

**Issue by SystemSigma** [rll/rllab#199]
Hello everybody!
I'm trying to train a 6 DOF robotic arm (UR5) to reach a 3D goal in its reachable space with DDPG, TRPO, ecc.
I've created my own MuJoCo asset and Gym environment to be launched in TF with GymEnv('env-id').
At each reset i sample uniformly the goal from a sphere centered with the arm base.
Since TRPO did not give any good results for 1M env steps of training with different batch sizes, i tried to sample the goal from a square centered on the initial gripper position, to reduce the space to be explored.

The environment is based on the `reacher.py` of Gym, with some adaptations for the reward function.
The goal is a mocap type body.
I set the terminal episode as the gripper of my arm come close less than 5cm to the goal.
Furthermore, i give penalties on arm autocollisions.
The env is below:

```
class UR5ReacherEnv(mujoco_env.MujocoEnv, utils.EzPickle):

    def __init__(self):
        utils.EzPickle.__init__(self)
        mujoco_env.MujocoEnv.__init__(self, 'UR5.xml', 4)

    def _step(self, a):
        self.dist_cost = np.linalg.norm(self.get_body_com("gripperpalm") - self.get_body_com("goal"))
        self.ctrl_cost = np.linalg.norm(a)

        if self.has_impact():
            self.impact_cost = 1
        else:
            self.impact_cost=0

        if self.dist_cost < 0.05:
            self.reward_final = 5
            done= True
        else:
            self.reward_final = 0
            done = False

        reward_dist = -self.dist_cost
        reward_ctrl = -1e-3 * self.ctrl_cost
        reward_impact = -self.impact_cost
        reward_final = self.reward_final


        reward = reward_dist + reward_ctrl + reward_impact + reward_final
        self.do_simulation(a, self.frame_skip)
        next_ob = self._get_obs()

        return next_ob, reward, done, dict(reward_dist=reward_dist,
                                           reward_ctrl=reward_ctrl,
                                           reward_impact=reward_impact,
                                           reward_final=reward_final)

    def viewer_setup(self):
        self.viewer.cam.trackbodyid = 2
        self.viewer.cam.distance = 2.5

    def reset_model(self):
        qpos = self.init_qpos \
               +self.np_random.uniform(low=-0.01, high=0.01, size=self.model.nq)
        qvel = self.init_qvel \
               +self.np_random.uniform(low=-0.1, high=0.1, size=self.model.nv)

        goal_yz = self.np_random.uniform(low=-.5, high=.5, size=2)

        self.goal=np.append(0.7, goal_yz)+[0, 0, 1]

        self.model.data.mocap_pos=self.goal

        self.set_state(qpos, qvel)

        self.dist = np.linalg.norm(self.get_body_com("gripperpalm") - self.get_body_com("goal"))
        return self._get_obs()

    def _get_obs(self):
        return self.state_vector()

    def has_impact(self):
        if  self.model.data.cfrc_ext.any()!=0:
            return True
        else: return False
```
Unfortunately I didn't get any improvements, as the average return is not improving monotonically as other Gym envs (has high fluctuations) Any suggestions?
Do I need more iterations to better explore the space or something is wrong with my reward?
Here is the log of my run with TRPO

[debug.log](https://github.com/rll/rllab/files/1495898/debug.log)

While the other exp hyperparameters are:

- baselineclass-linear
- gaelambda-0-97
- ksb-0
- policy hidden nonlinearity-relu
- policy hidden sizes-400x300
- policy output nonlinearity-tanh
- ss-0-01

Thanks to everyone that will help :)

**Comment by AhmedMagdyHendawy** [comment]
did u solve this problem ?
